rfcs/20200817-jit-code-generation-for-aarch64/README.md

```diff
@@ -0,0 +1,256 @@
+(Fig. Indirect method). Since we can reuse jit_\*.(cpp|hpp) for x64 with just 
+a few changes for AArch64, and we can reduce our development effort for 
+AArch64 to a great extent. We call this as "Indirect method". So **this RFC 
+also introduces Xbyak_translator_aarch64 and 
+[xed](https://github.com/fujitsu/xed "xed") for AArch64.**
+
+![Indirect method](indirect_method.png)
+
+
+As a third method, it is also possible that we use x64's jit_\*.(cpp|hpp) 
+files and their translation to AArch64 as a baseline, and partially rewrite 
+them from Xbyak functions to Xbyak_aarch64 ones to achieve better performance 
+for AArch64. We call this method as "Mixed-method".
+
+In summary, this RFC allows us to generate JIT code for AArch64 in the 
+following three ways.
+- Direct method:Implement JIT code generation from scratch using Xbyak_aarch64
+- Indirect method:Generate JIT code for AArch64 from jit_\*.(cpp|hpp) 
+for x64 by Xbyak_translator_aarch64
+- Mixed method:Use jit_\*.(cpp|hpp) files for x64 as a baseline, tune them 
+partially with Xbyak_aarch64
+
+We applied these three methods to oneDNN code in v0.21 and confirms that 
+the JIT code is generated on AArch64, it passes tests/gtests/test_\*, 
+and it accelerates oneDNN processing as shows in the slide below.
+The source code is disclosed as 
+[https://github.com/fujitsu/dnnl_aarch64](https://github.com/fujitsu/dnnl_aarch64).
+
+
+![Performance evaluation of JIT](performance_evaluation.png)
+
+
+## Related RFCs and Issues
+
+- [https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200401-cpu-dir-structure](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200401-cpu-dir-structure)
+- [https://github.com/oneapi-src/oneDNN/pull/795](https://github.com/oneapi-src/oneDNN/pull/795)
+- [https://github.com/oneapi-src/oneDNN/issues/804](https://github.com/oneapi-src/oneDNN/issues/804)
+
+
+## Proposal
+
+- Adding xbyak_translator_aarch64 directories to src/cpu/aarch64
+- Adding a dependency on xed at the build stage
+- Adding primitive entry to src/cpu/cpu_\*_list.cpp
+
+### Adding xbyak_translator_aarch64 directories to src/cpu/aarch64
+
+The following tree shows the directory structure of xbyak_translator_aarch64. 
+Some subdirectories are abbreviated from the tree. Xbyak_translator_aarch64 
+contains Xbyak and Xbyak_aarch64. Xbyak_translator_aarch64 is built by forking 
+[Xbyak](https://github.com/herumi/xbyak), and the directory structure follows 
+that of xbyak. The difference between the original `Xbyak` and 
+`Xbyak_translator_aarch64` can be seen by comparing 
+https://github.com/fujitsu/xbyak_translator_aarch64/tree/master and 
+https://github.com/fujitsu/xbyak_translator_aarch64/tree/fjdev. 
+
+Comparing to the original `Xbyak`, call of `decodeAndTransToAArch64` function 
+is added to each function of `xbyak_mnemonic.h`. `decodeAndTransToAArch64` 
+disassembles x64 instruction, which is generated by xbyak, by `xed`, and call 
```

Comment:
Out of curiosity, why do you need to encode the instruction with Xbyak first, to decode it then with `xed`? Doesn't Xbyak contain all necessary information already?

My guess is that it is probably done for the convenience, because otherwise you would need to embed not a simple `decodeAndTransToAArch64` in each Xbyak mnemonic, but actual `TransThisInstrToAArch64()` and `TransThatInstrToAArch64()`. On the other hand that would reduce the number of dependencies.

Reply:
You make a good point. As you pointed out, we chose this configuration because it is easier to implement and less buggy.

Reply:
I see. Thanks for clarification!